js/execution.js

```diff
@@ -0,0 +1,249 @@
+     * current scope (thus, "the stack")
+     **/
+    function Context(instruction, scope) {
+        this.currentInstruction = instruction;
+        this.scope = scope;
+    }
+
+    /**
+     * Runes the next instruction, updating the instruction pointer and scope
+     * as appropriate.
+     *
+     * Returns: true if this context has more instructions to execute.
+     */
+    Context.prototype.doNext = function next() {
+        assert(this.currentInstruction !== null, 'current instruction undefined.');
+
+        /* The scope is mutable so... it gets mutated. */
+        this.currentInstruction.run(this.scope);
+        this.currentInstruction = this.next();
+
+        return this.currentInstruction !== null;
+    };
+
+    /**
+     * Get the next instruction in this context.
+     */
+    Context.prototype.next = function next() {
+        var nextInstruction;
+
+        /* FIXME: TERRIBLE SEPARATION OF CONCERNS AND ALSO WRONG. */
+        nextInstruction = this.currentInstruction.nextElementSibling;
+
+        assert(nextInstruction === null || typeof nextInstruction.run === 'function',
+               'Block does not have a callable property `run`');
+
+        return nextInstruction;
+    };
+
+    /**
+     * Creates the root context -- that is, the context from which all other
+     * contexts originate from.
+     */
+    Context.createRootContext = function createRootContext() {
+        var globalScope = {},
+            /* FIXME: THIS DOM STUFF DOES NOT BELONG HERE! */
+            firstInstruction = dom.find('wb-workspace > wb-contains > *');
+
+        return new Context(firstInstruction, globalScope);
+    };
+
+
+    /**
+     * Encapsulates the execution of a Waterbear program. There may be several
+     * "threads", so this encompasses all of them. ALL OF THEM.
+     */
+    function Process() {
+        var started = false;
+
+        /* Set some essential state. */
+        this.contexts = [];
+        this.currentContext = null;
+        this.paused = false;
+
+        /* Disable breakpoints. */
+        this.shouldBreak = false;
+        /* Run as quickly as possible. */
+        this.delay = 0;
+        this.nextTimeout = null;
+
+        this.doNextStep = this.nextStep.bind(this);
+
+        Object.defineProperties(this, {
+            /* `started` is trapped in this closure, thus it can only be
+             * changed in this constructor, or any function defined within this
+             * constructor. */
+            started: {
+                get: function () { return started; },
+                enumerable: true
+            },
+            /* There's no going back once this function gets called! */
+            setStarted: {
+                value: function () { started = true; },
+                enumerable: false
+            }
+        });
+    }
+
+    /**
+     * Starts (asynchronous!) execution from scratch. Can only be called once.
+     */
```

Comment:
Good point. Yeah, all of this non-sense here is so that each `WaterbearProcess` is a use-once only object. Every time start is pressed, it actually _tosses out_ any existing `WaterbearProcess` instance, and instantiates a new one. I decided this was a good idea because the process instance stores _a lot of state_, and this way I can throw it all out all in one go and start from scratch every time `start` is pressed. As noted in #998, we still have an issue with storing state from previous runs... somewhere-- and I wanted to avoid that altogether.

But I agree, it does look weird.